pkg/process/process.go

```diff
@@ -0,0 +1,234 @@
+func NewTree(ctx context.Context, resetDuration time.Duration) *Tree {
+	t := &Tree{
+		tree:        make(map[key]*process),
+		disappeared: make(map[key]struct{}),
+		mtx:         &sync.RWMutex{},
+	}
+	go func() {
+		// This is a very naive pruning implementation. It will be improved in the future.
+		ticker := time.NewTicker(resetDuration)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				// Prune disappeared processes.
+				t.mtx.Lock()
+				for k := range t.disappeared {
+					delete(t.tree, k)
+				}
+				t.mtx.Unlock()
+				t.disappeared = make(map[key]struct{})
+
+				t.mtx.RLock()
+				for k, p := range t.tree {
+					proc, err := procfs.NewProc(p.PID)
+					if err != nil {
+						// Will be pruned in the next iteration.
+						t.disappeared[k] = struct{}{}
+						continue
+					}
+					stat, err := proc.Stat()
+					if err != nil {
+						// Will be pruned in the next iteration.
+						t.disappeared[k] = struct{}{}
+						continue
+					}
+					// It is possible that the process has been reused. Update the process.
+					if p.starttime != stat.Starttime {
+						t.reused[k] = &process{
+							Proc:      proc,
+							tree:      t,
+							parent:    stat.PPID,
+							starttime: stat.Starttime,
+						}
+						continue
+					}
+				}
+				t.mtx.RUnlock()
+
+				// Update the process tree.
+				t.mtx.Lock()
+				for k, p := range t.reused {
+					t.tree[k] = p
+				}
+				t.mtx.Unlock()
+				t.reused = make(map[key]*process)
+			}
+		}
+	}()
+	return t
+}
+
+// FindAllAncestorProcessIDsInSameCgroup returns all ancestor process IDs for a given PID in the same cgroup.
+func (t *Tree) FindAllAncestorProcessIDsInSameCgroup(pid int) ([]int, error) {
+	// Fast path. Find the process if it exists in the process tree.
+	k := key{pid: pid}
+	t.mtx.RLock()
+	if p, ok := t.tree[k]; ok {
+		t.mtx.RUnlock()
+		// Process could have been already terminated.
+		// And this could be a problem if we haven't updated the process tree yet.
+		proc, err := procfs.NewProc(pid)
+		if err != nil {
+			return nil, err
+		}
+		stat, err := proc.Stat()
+		if err != nil {
+			return nil, err
+		}
+		if p.starttime == stat.Starttime {
+			return findAncestorPIDsInSameCgroup(p)
+		}
+		// Same PID but different start time. PID has been reused.
+	}
+	t.mtx.RUnlock()
+
+	// Update the process tree.
+	if err := t.populate(); err != nil {
+		return nil, fmt.Errorf("failed to populate process tree: %w", err)
+	}
+
+	t.mtx.RLock()
+	// Tree is updated. Try to find the process again.
+	p, ok := t.tree[k]
+	if !ok {
+		t.mtx.RUnlock()
+		return nil, fmt.Errorf("failed to find the process for PID %d", pid)
+	}
+	t.mtx.RUnlock()
+	return findAncestorPIDsInSameCgroup(p)
+}
+
+func (t *Tree) populate() error {
+	procs, err := procfs.AllProcs()
+	if err != nil {
+		return fmt.Errorf("failed to get all processes: %w", err)
+	}
+
+	t.mtx.Lock()
```

Comment:
It's rather unpredictable how long we'll hold this lock for, so I think we should only obtain a read lock here and return the changes that need to happen to the map, and then only obtain the lock for the short moment where we populate the map with the new changes.

I would also suggest that this function doesn't actually obtain the lock at all, and rather just returns the difference. It's easier to reason about locking behavior when all the locking/unlocking happens in one function.

Another benefit of moving the write-lock to the caller of this function is that the subsequent read that we currently hold a read lock for can just be done as part of the write-lock.